File: frontend/src/types/recurring.ts
export type RecurringFrequency = 'daily' | 'weekly' | 'monthly' | 'yearly';
export type RecurringStatus = 'active' | 'paused' | 'cancelled';

export interface RecurringRule {
  id: string;
  user_id: string;
  account_id: string;
  category_id: string | null;
  budget_id: string | null;
  name: string;
  type: 'income' | 'expense';
  amount: string;
  currency: string;
  frequency: RecurringFrequency;
  start_date: string;
  end_date: string | null;
  next_occurrence: string;
  is_subscription: boolean;
  status: RecurringStatus;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface RecurringRuleCreate {
  account_id: string;
  category_id?: string | null;
  budget_id?: string | null;
  name: string;
  type: 'income' | 'expense';
  amount: string;
  currency: string;
  frequency: RecurringFrequency;
  start_date: string;
  end_date?: string | null;
  is_subscription?: boolean;
  notes?: string | null;
}

export interface RecurringRuleUpdate {
  name?: string;
  amount?: string;
  category_id?: string | null;
  budget_id?: string | null;
  end_date?: string | null;
  is_subscription?: boolean;
  status?: RecurringStatus;
  notes?: string | null;
}

export interface UpcomingSubscription {
  id: string;
  name: string;
  amount: string;
  currency: string;
  next_occurrence: string;
  frequency: RecurringFrequency;
  days_until: number;
}

File: backend/app/schemas/recurring.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel


class RecurringRuleCreate(BaseModel):
    account_id: uuid.UUID
    category_id: uuid.UUID | None = None
    budget_id: uuid.UUID | None = None
    name: str
    type: str  # income, expense
    amount: Decimal
    currency: str
    frequency: str  # daily, weekly, monthly, yearly
    start_date: date
    end_date: date | None = None
    is_subscription: bool = False
    notes: str | None = None


class RecurringRuleUpdate(BaseModel):
    name: str | None = None
    amount: Decimal | None = None
    category_id: uuid.UUID | None = None
    budget_id: uuid.UUID | None = None
    end_date: date | None = None
    is_subscription: bool | None = None
    status: str | None = None
    notes: str | None = None


class RecurringRuleResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    account_id: uuid.UUID
    category_id: uuid.UUID | None
    budget_id: uuid.UUID | None
    name: str
    type: str
    amount: Decimal
    currency: str
    frequency: str
    start_date: date
    end_date: date | None
    next_occurrence: date
    is_subscription: bool
    status: str
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class UpcomingSubscription(BaseModel):
    id: uuid.UUID
    name: str
    amount: Decimal
    currency: str
    next_occurrence: date
    frequency: str
    days_until: int
